component/common/loki/client/internal/marker_handler.go

```diff
@@ -0,0 +1,178 @@
+		quit:         make(chan struct{}),
+	}
+
+	// Load the last marked segment from disk (if it exists).
+	if lastSegment := mh.markerFileHandler.LastMarkedSegment(); lastSegment >= 0 {
+		mh.lastMarkedSegment = lastSegment
+	}
+
+	mh.wg.Add(1)
+	go mh.runUpdatePendingData()
+
+	return mh
+}
+
+func (mh *markerHandler) LastMarkedSegment() int {
+	return mh.markerFileHandler.LastMarkedSegment()
+}
+
+func (mh *markerHandler) UpdateReceivedData(segmentId, dataCount int) {
+	mh.dataIOUpdate <- dataUpdate{
+		segmentId: segmentId,
+		dataCount: dataCount,
+	}
+}
+
+func (mh *markerHandler) UpdateSentData(segmentId, dataCount int) {
+	mh.dataIOUpdate <- dataUpdate{
+		segmentId: segmentId,
+		dataCount: -1 * dataCount,
+	}
+}
+
+type countDataItem struct {
+	count      int
+	lastUpdate time.Time
+}
+
+type processDataItem struct {
+	segment    int
+	count      int
+	lastUpdate time.Time
+}
+
+// runUpdatePendingData is assumed to run in a separate routine, asynchronously keeping track of how much data each WAL
+// segment the Watcher reads from, has left to send. When a segment reaches zero, it means that is has been consumed,
+// and the highest numbered one is marked as consumed.
+func (mh *markerHandler) runUpdatePendingData() {
+	defer mh.wg.Done()
+
+	segmentDataCount := make(map[int]countDataItem)
+
+	for {
+		select {
+		case <-mh.quit:
+			return
+		case update := <-mh.dataIOUpdate:
+			if di, ok := segmentDataCount[update.segmentId]; ok {
+				di.lastUpdate = time.Now()
+				di.count += update.dataCount
+			} else {
+				segmentDataCount[update.segmentId] = countDataItem{
+					count:      update.dataCount,
+					lastUpdate: time.Now(),
+				}
+			}
+		}
+
+		// TODO: Add rate limit here to not trigger this N log N algo on every update
+
+		markableSegment := FindMarkableSegment(segmentDataCount, time.Hour)
+		if markableSegment > mh.lastMarkedSegment {
+			mh.markerFileHandler.MarkSegment(markableSegment)
+			mh.lastMarkedSegment = markableSegment
+		}
+	}
+}
+
+func (mh *markerHandler) Stop() {
+	mh.quit <- struct{}{}
+	mh.wg.Wait()
+}
+
+// FindMarkableSegment finds, given the summary of data updates received, and a threshold on how much time can pass for
+// a segment that hasn't received updates to be considered as "live", the segment that should be marked as last consumed.
+// The algorithm will find the highest numbered segment that is considered as "consumed", with its all predecessors
+// "consumed" as well.
+//
+// A consumed segment is one with data count of zero, meaning that there's no data left in flight for it, or it hasn't
+// received any updates for tooOldThreshold time.
+//
+// Also, while reviewing the data items in segmentDataCount, those who are consumed will be deleted to clean up space.
+//
+// This algorithm runs in O(N log N), being N the size of segmentDataCount, and allocates O(N) memory.
+func FindMarkableSegment(segmentDataCount map[int]countDataItem, tooOldThreshold time.Duration) int {
```

Comment:
@ptodev one thing we can do to improve this after this first version, is use for `segmentDataCount` a data structure that is:
- a map
- support iteration over it's keys

I imagine something like a tree-based map that allows an in-order walk over it's items would work. But I repeat, for later and to minimize the N log N, which might not be even needed, since my idea is to run the mark segment not for each update we receive, but only every so much time (like limited by a ticker).

Reply:
Thank you, the new algorithm makes sense. I'm just not sure whether the `tooOldThreshold` should be configurable. Atm it is hardcoded to one hour - I wonder why this value was chosen? It'd be good to clarify it in a comment please.

Reply:
Oh yhea forgot to parametrize the value, added `time.Hour` for testing.

Reply:
Used [`MaxSegmentAge`](https://github.com/grafana/agent/blob/main/component/common/loki/wal/config.go#L31) to configure the tooOldThreshold